core/logger/audit/audit_logger.go

```diff
@@ -0,0 +1,312 @@
+	}
+
+	if _, err := models.ParseURL(forwardToUrl); err != nil {
+		return AuditLoggerConfig{}, errors.Errorf("forwardToURL value is not a valid URL")
+	}
+
+	environment := "production"
+	if isDev {
+		environment = "develop"
+	}
+
+	// Split and prepare optional service client headers from env variable
+	headers := []ServiceHeader{}
+	if encodedHeaders != "" {
+		headerLines := strings.Split(encodedHeaders, "\\")
+		for _, header := range headerLines {
+			keyValue := strings.Split(header, "||")
+			if len(keyValue) != 2 {
+				return AuditLoggerConfig{}, errors.Errorf("Invalid headers provided for the audit logger. Value, single pair split on || required, got: %s", keyValue)
+			}
+			headers = append(headers, ServiceHeader{
+				Header: keyValue[0],
+				Value:  keyValue[1],
+			})
+		}
+	}
+
+	return AuditLoggerConfig{
+		ForwardToUrl:   &forwardToUrl,
+		Environment:    &environment,
+		JsonWrapperKey: &jsonWrapperKey,
+		Headers:        headers,
+	}, nil
+}
+
+type AuditLoggerService struct {
+	logger          logger.Logger            // The standard logger configured in the node
+	enabled         bool                     // Whether the audit logger is enabled or not
+	forwardToUrl    string                   // Location we are going to send logs to
+	headers         []ServiceHeader          // Headers to be sent along with logs for identification/authentication
+	jsonWrapperKey  string                   // Wrap audit data as a map under this key if present
+	environmentName string                   // Decorate the environment this is coming from
+	hostname        string                   // The self-reported hostname of the machine
+	localIP         string                   // A non-loopback IP address as reported by the machine
+	loggingClient   HTTPAuditLoggerInterface // Abstract type for sending logs onward
+
+	loggingChannel chan wrappedAuditLog
+	ctx            context.Context
+	cancel         context.CancelFunc
+	chDone         chan struct{}
+}
+
+// Configurable headers to include in POST to log service
+type ServiceHeader struct {
+	Header string
+	Value  string
+}
+
+type wrappedAuditLog struct {
+	eventID EventID
+	data    Data
+}
+
+// NewAuditLogger returns a buffer push system that ingests audit log events and
+// asynchronously pushes them up to an HTTP log service.
+// Parses and validates the AUDIT_LOGS_* environment values and returns an enabled
+// AuditLogger instance. If the environment variables are not set, the logger
+// is disabled and short circuits execution via enabled flag.
+func NewAuditLogger(logger logger.Logger, config *AuditLoggerConfig) (AuditLogger, error) {
+	if config == nil {
+		return &AuditLoggerService{}, errors.Errorf("Audit Log initialization error - no configuration")
+	}
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		return &AuditLoggerService{}, errors.Errorf("Audit Log initialization error - unable to get hostname: %s", err)
```

Comment:
This is pretty unusual. IMO using the idiomatic error _or_ result pattern will makes things much more readable at the callsite.
```suggestion
		return nil, errors.Errorf("Audit Log initialization error - unable to get hostname: %s", err)
```

Reply:
Currently we use this to avoid code duplication implementing a null logger.

Reply:
This is a non-idiomatic, confusing way to implement a null logger in Go and within this code base. Please follow the traditional convention of the result being unused when error is present. The caller can explicitly create a null logger if they still need one.